File: backend/src/main/java/br/ufpr/sept/so2/modules/presenca/infrastructure/persistence/EventoJpaAdapter.java
package br.ufpr.sept.so2.modules.presenca.infrastructure.persistence;

import br.ufpr.sept.so2.modules.presenca.application.ports.EventoRepository;
import br.ufpr.sept.so2.modules.presenca.domain.Evento;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Component;

import java.time.OffsetDateTime;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Component
public class EventoJpaAdapter implements EventoRepository {

    private final EventoJpaRepository jpaRepository;

    public EventoJpaAdapter(EventoJpaRepository jpaRepository) {
        this.jpaRepository = jpaRepository;
    }

    @Override
    public Evento save(Evento evento) {
        EventoJpaEntity entity = jpaRepository.findById(evento.getId())
                .orElseGet(() -> EventoJpaEntity.fromDomain(evento));
        entity.merge(evento);
        return jpaRepository.save(entity).toDomain();
    }

    @Override
    public Optional<Evento> findById(UUID id) {
        return jpaRepository.findById(id).map(EventoJpaEntity::toDomain);
    }

    @Override
    public Page<Evento> findAbertosParaAluno(OffsetDateTime agora, Pageable pageable) {
        return jpaRepository.findAbertosParaAluno(agora, pageable).map(EventoJpaEntity::toDomain);
    }

    @Override
    public List<Evento> findEmAndamentoComJanelaNoDia(OffsetDateTime inicioDia, OffsetDateTime fimDia) {
        return jpaRepository.findEmAndamentoComJanelaNoDia(inicioDia, fimDia).stream()
                .map(EventoJpaEntity::toDomain)
                .toList();
    }

    @Override
    public Page<Evento> findByAnfitriao(UUID anfitriaoId, Pageable pageable) {
        return jpaRepository.findByIdAnfitriao(anfitriaoId, pageable).map(EventoJpaEntity::toDomain);
    }

    @Override
    public boolean existsByTitulo(String titulo) {
        return jpaRepository.existsByTituloIgnoreCase(titulo);
    }

    @Override
    public Optional<Evento> findByTitulo(String titulo) {
        return jpaRepository.findByTituloIgnoreCase(titulo).map(EventoJpaEntity::toDomain);
    }
}

File: backend/src/main/kotlin/br/ufpr/sept/so2/modules/presenca/infrastructure/persistence/EventoJpaAdapter.kt
package br.ufpr.sept.so2.modules.presenca.infrastructure.persistence

import br.ufpr.sept.so2.modules.presenca.application.ports.EventoRepository
import br.ufpr.sept.so2.modules.presenca.domain.Evento
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Component
import java.time.OffsetDateTime
import java.util.Optional
import java.util.UUID

@Component
class EventoJpaAdapter(
    private val jpaRepository: EventoJpaRepository,
) : EventoRepository {

    override fun save(evento: Evento): Evento {
        val entity = jpaRepository.findById(evento.id)
            .orElseGet { EventoJpaEntity.fromDomain(evento) }
        entity.merge(evento)
        return jpaRepository.save(entity).toDomain()
    }

    override fun findById(id: UUID): Optional<Evento> = jpaRepository.findById(id).map { it.toDomain() }

    override fun findAbertosParaAluno(agora: OffsetDateTime, pageable: Pageable): Page<Evento> =
        jpaRepository.findAbertosParaAluno(agora, pageable).map { it.toDomain() }

    override fun findEmAndamentoComJanelaNoDia(inicioDia: OffsetDateTime, fimDia: OffsetDateTime): List<Evento> =
        jpaRepository.findEmAndamentoComJanelaNoDia(inicioDia, fimDia).map { it.toDomain() }

    override fun findByAnfitriao(anfitriaoId: UUID, pageable: Pageable): Page<Evento> =
        jpaRepository.findByIdAnfitriao(anfitriaoId, pageable).map { it.toDomain() }

    override fun existsByTitulo(titulo: String): Boolean = jpaRepository.existsByTituloIgnoreCase(titulo)

    override fun findByTitulo(titulo: String): Optional<Evento> =
        jpaRepository.findByTituloIgnoreCase(titulo).map { it.toDomain() }
}
